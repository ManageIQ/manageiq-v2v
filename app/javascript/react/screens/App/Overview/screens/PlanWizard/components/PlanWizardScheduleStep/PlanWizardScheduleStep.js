import React from 'react';
import PropTypes from 'prop-types';
import { Field, reduxForm } from 'redux-form';
import { Form, Icon } from 'patternfly-react';
import { FormField } from '../../../../../common/forms/FormField';

export const PlanWizardScheduleStep = ({
  targetProvider,
  migration_plan_type_radio,
  migration_plan_choice_radio,
  enableWarmMigration
}) => (
  <Form className="form-vertical">
    <Field
      name="migration_plan_type_radio"
      component={FormField}
      type="radio"
      label={__('Migration Type')}
      inline_label={false}
      controlWidth={12}
      info={
        enableWarmMigration
          ? null
          : __("Warm migration is not possible because one or more selected VMs has disks that can't be pre-copied.")
      }
      infoIconName="warning-triangle-o"
      options={[
        {
          id: 'migration_type_cold',
          name: __('Cold Migration - Source VMs are shut down while all of the VM data is migrated.')
        },
        {
          id: 'migration_type_warm',
          name: __('Warm Migration - VM data is iteratively pre-copied. A final, cutover migration is scheduled and run later.'), // prettier-ignore
          disabled: !enableWarmMigration
        }
      ]}
    />
    {migration_plan_type_radio === 'migration_type_cold' && (
      <Field
        name="migration_plan_choice_radio"
        component={FormField}
        type="radio"
        label={__('Schedule')}
        inline_label={false}
        style={{ paddingTop: '100px' }}
        options={[
          {
            name: __('Save migration plan to run later'),
            id: 'migration_plan_later'
          },
          {
            name: __('Start migration immediately'),
            id: 'migration_plan_now'
          }
        ]}
        help={
          targetProvider === 'openstack' && migration_plan_choice_radio === 'migration_plan_now' ? (
            <div className="plan-wizard-vm-power-warning">
              <Icon type="pf" name="warning-triangle-o" />
              <p>
                {__('VMs must be powered on in order to migrate.')}
                <br />
                {__('Ensure that all VMs in the Migration Plan are powered on before selecting Create.')}
              </p>
            </div>
          ) : null
        }
      />
    )}
    {migration_plan_type_radio === 'migration_type_warm' && (
      <Form.HelpBlock style={{ paddingLeft: '22px', paddingTop: '100px' }}>
        {__('Pre-copying of the VM data will begin immediately after plan creation. You can schedule and run final cutover migration later.') /* prettier-ignore */}
      </Form.HelpBlock>
    )}
  </Form>
);

PlanWizardScheduleStep.propTypes = {
  targetProvider: PropTypes.string,
  migration_plan_type_radio: PropTypes.string,
  migration_plan_choice_radio: PropTypes.string,
  enableWarmMigration: PropTypes.bool
};

export default reduxForm({
  form: 'planWizardScheduleStep',
  destroyOnUnmount: false
})(PlanWizardScheduleStep);
