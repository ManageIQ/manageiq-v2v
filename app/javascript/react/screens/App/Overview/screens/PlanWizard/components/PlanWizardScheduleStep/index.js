import { connect } from 'react-redux';
import PlanWizardScheduleStep from './PlanWizardScheduleStep';
import { getCurrentTargetProvider } from '../../PlanWizardSelectors';

const mapStateToProps = ({ overview: { transformationMappings }, form }) => ({
  targetProvider: getCurrentTargetProvider(form, transformationMappings),
  migration_plan_choice_radio:
    form.planWizardScheduleStep &&
    form.planWizardScheduleStep.values &&
    form.planWizardScheduleStep.values.migration_plan_choice_radio,
  migration_plan_type_radio:
    form.planWizardScheduleStep &&
    form.planWizardScheduleStep.values &&
    form.planWizardScheduleStep.values.migration_plan_type_radio,
  initialValues: {
    migration_plan_choice_radio: 'migration_plan_later',
    migration_plan_type_radio: 'migration_type_cold'
  }
});

export default connect(mapStateToProps)(PlanWizardScheduleStep);
