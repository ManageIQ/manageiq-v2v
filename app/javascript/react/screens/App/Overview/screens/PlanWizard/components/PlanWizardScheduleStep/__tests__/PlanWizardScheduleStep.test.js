import React from 'react';
import { shallow } from 'enzyme';
import { PlanWizardScheduleStep } from '../PlanWizardScheduleStep';

describe('Plan wizard schedule step', () => {
  it('renders with a warning if migrating an OSP plan immediately', () => {
    const component = shallow(
      <PlanWizardScheduleStep
        targetProvider="openstack"
        migration_plan_type_radio="migration_type_cold"
        migration_plan_choice_radio="migration_plan_now"
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders with no warning if migrating an OSP plan later', () => {
    const component = shallow(
      <PlanWizardScheduleStep
        targetProvider="openstack"
        migration_plan_type_radio="migration_type_cold"
        migration_plan_choice_radio="migration_plan_later"
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders with a info text when warm migration selected', () => {
    const component = shallow(
      <PlanWizardScheduleStep targetProvider="openstack" migration_plan_type_radio="migration_type_warm" />
    );
    expect(component).toMatchSnapshot();
  });
});
