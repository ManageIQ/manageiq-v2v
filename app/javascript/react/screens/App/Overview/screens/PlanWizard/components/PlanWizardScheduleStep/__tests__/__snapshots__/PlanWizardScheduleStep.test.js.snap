// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan wizard schedule step renders with a warning if migrating an OSP plan immediately 1`] = `
<Form
  bsClass="form"
  className="form-vertical"
  componentClass="form"
  horizontal={false}
  inline={false}
>
  <Field
    component={[Function]}
    controlWidth={12}
    inline_label={false}
    label="Migration Type"
    name="migration_plan_type_radio"
    options={
      Array [
        Object {
          "id": "migration_type_cold",
          "name": "Cold Migration - Source VMs are shut down while all of the VM data is migrated.",
        },
        Object {
          "id": "migration_type_warm",
          "name": "Warm Migration - VM data is iteratively pre-copied. A final, cutover migration is scheduled and run later.",
        },
      ]
    }
    type="radio"
  />
  <Field
    component={[Function]}
    help={
      <div
        className="plan-wizard-vm-power-warning"
      >
        <Icon
          name="warning-triangle-o"
          type="pf"
        />
        <p>
          VMs must be powered on in order to migrate.
          <br />
          Ensure that all VMs in the Migration Plan are powered on before selecting Create.
        </p>
      </div>
    }
    inline_label={false}
    label="Schedule"
    name="migration_plan_choice_radio"
    options={
      Array [
        Object {
          "id": "migration_plan_later",
          "name": "Save migration plan to run later",
        },
        Object {
          "id": "migration_plan_now",
          "name": "Start migration immediately",
        },
      ]
    }
    style={
      Object {
        "paddingTop": "100px",
      }
    }
    type="radio"
  />
</Form>
`;

exports[`Plan wizard schedule step renders with no warning if migrating an OSP plan later 1`] = `
<Form
  bsClass="form"
  className="form-vertical"
  componentClass="form"
  horizontal={false}
  inline={false}
>
  <Field
    component={[Function]}
    controlWidth={12}
    inline_label={false}
    label="Migration Type"
    name="migration_plan_type_radio"
    options={
      Array [
        Object {
          "id": "migration_type_cold",
          "name": "Cold Migration - Source VMs are shut down while all of the VM data is migrated.",
        },
        Object {
          "id": "migration_type_warm",
          "name": "Warm Migration - VM data is iteratively pre-copied. A final, cutover migration is scheduled and run later.",
        },
      ]
    }
    type="radio"
  />
  <Field
    component={[Function]}
    help={null}
    inline_label={false}
    label="Schedule"
    name="migration_plan_choice_radio"
    options={
      Array [
        Object {
          "id": "migration_plan_later",
          "name": "Save migration plan to run later",
        },
        Object {
          "id": "migration_plan_now",
          "name": "Start migration immediately",
        },
      ]
    }
    style={
      Object {
        "paddingTop": "100px",
      }
    }
    type="radio"
  />
</Form>
`;
